src/build/app/mod.rs

```diff
@@ -2127,25 +2126,12 @@ impl<'help> App<'help> {
         T: Into<OsString> + Clone,
     {
         let mut it = Input::from(itr.into_iter());
-        // Get the name of the program (argument 1 of env::args()) and determine the
-        // actual file
-        // that was used to execute the program. This is because a program called
-        // ./target/release/my_prog -a
-        // will have two arguments, './target/release/my_prog', '-a' but we don't want
-        // to display
-        // the full path when displaying help messages and such
         if !self.settings.is_set(AppSettings::NoBinaryName) {
-            if let Some((name, _)) = it.next() {
-                let p = Path::new(name);
-
-                if let Some(f) = p.file_name() {
-                    if let Some(s) = f.to_str() {
-                        if self.bin_name.is_none() {
-                            self.bin_name = Some(s.to_owned());
-                        }
-                    }
-                }
+            if self.bin_name.is_none() {
+                self.bin_name = Some(self.name.to_owned());
```

Comment:
This is an unergonomic design. We want to have a good default for having the binary name be actually the binary name. And there is `NoBinaryName` setting if people don't want it.

You are trying to solve the issue that `print_help` do not use `bin_name`, right? Maybe a different approach would be better.